### ISSUE TITLE
neuroevolution algorithms | 大トロ

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#1382]
neuroevolution algorithms | 大トロ<br>
<br>
https://ift.tt/2uESa7x<br>
<br>
<br>
<br>

<div>
<section><p>I haven’t had much look tackling the pendulum balancing problem with Q-learning.  It seems that there are many states in the system, and the output (motor speed) should really be a continuous variable, that it didn’t work well, with the Q-learner that spits out discrete speeds, or even generate faster, stay, slower discretised states.</p>

<p>The problem is that by using conventional gradient descent methods common in deep learning, we are trying to ‘solve’ the weights of a neural network in such a way that the network learns how the transfer function of a system works, ie, predict the output of a system given some input, rather than trying to find a strategy.  While this may certainly be useful in modelling how physical systems like the pendulum works, in the sense that it predicts the orientation/velocities at the next time step given the variables of the current state, it may not be able to come up with a strategy to how to get to a certain desired state, especially when the current state (say the pendulum is inverted completely) is so far away from the desired state 180 degrees away.  Not to mention that it’s got to stay up there once it swings all the way up.</p>

<p>I’ve began to question the idea to use SGD / back-propagation to train networks to achieve tasks, as there may be too many local minimums.  In addition, I think the Q-learning algorithm described in the atari paper essentially hacks the memory element into a feed-forward network system.  The main contribution from that paper is essentially the use of convolutional networks processing video game images, and having the back-prop work all the way from expected reward down to features extracted during convolution, so it can work very fast on GPUs.  I am thinking the more correct approach is to bite the bullet and use recurrent neural networks that can incorporate feedback and memory elements for task-based training.  This may make it very difficult to train the recurrent networks using SGD / backprop.  We don’t even really know what desired output we want the network to generate given the current input state.</p>

<div id="attachment_274" class="wp-caption alignnone">
  <a href="http://blog.otoro.net/wp-content/uploads/sites/2/2015/02/cne_basic.png"><img src="http://blog.otoro.net/wp-content/uploads/sites/2/2015/02/cne_basic.png"></a>

  <p class="wp-caption-text">
    source: wikipedia
  </p>
</div>

<p>During the evenings, I have been reading up on using genetic algorithm approaches to train neural networks.  The field seems to be studied in great depth at the <a title="NN lab at U of Texas" href="http://nn.cs.utexas.edu/" target="_blank">University of Texas</a>, and the <a title="IDSIA" href="http://idsia.ch/" target="_blank">Swiss AI Lab</a>‘s work on RNNs, and strangely I haven’t heard about much of this work at <a title="stanford deep learning tutorial" href="http://ufldl.stanford.edu/wiki/index.php/UFLDL_Tutorial" target="_blank">Stanford</a> or <a title="U of T" href="http://www.cs.toronto.edu/~hinton/" target="_blank">University of Toronto</a> (my <a title="engsci" href="http://engsci.utoronto.ca/" target="_blank">alma mater</a>!) or <a title="NYU ai lab" href="http://www.cs.nyu.edu/~yann/" target="_blank">NYU</a> so much with all the buzz about deep learning.</p>

<p>A good introduction to the field I found was on these <a title="slides on neuroevolution" href="http://nn.cs.utexas.edu/?neuroevolution-tutorial-ijcnn2013" target="_blank">slides</a> at the U of Texas NN lab.  Digging through some older thesis publications, I found John Gomez’s <a title="ESP algorithm neuroevolution" href="http://www.cs.utexas.edu/users/nn/downloads/papers/gomez.phdtr03.pdf" target="_blank">thesis</a> extremely helpful and well written, and as someone who has not implemented a Genetic Algorithm before, the pseudocode in the thesis put things to light on how this stuff works.</p>

<p>I realised that these methods may be superior to Q-learning or backprop based to search for a strategy that requires a sequence of multiple actions.  Even the simplest neuroevolution algorithm (‘conventional neuroevolution’), the plain vanilla algorithm with no extra bells and whistles, maybe able to solve many interesting problems.  What this simple algorithm does is:</p>

<ol>
  <li>
    <p>Take a certain neural network architecture, feed forward, or even recurrent, and make N number (say 100) of these neural nets each randomised with different weights.</p>
  </li>
  <li>
    <p>Try to accomplish the task at hand (balance the pendulum), for each of the 100 networks, and then score how well each performs the task (assign the score of say the average angle squared during the life of the task, where zero degree is upright)</p>
  </li>
  <li>
    <p>Sort the networks by scores achieved, and keep the top 20 networks, and throw the rest away.  For each network, dump all the weights and biases into a one-dimensional array of floats, in a consistent orderly way, and call that the chromosome of the network.</p>
  </li>
  <li>
    <p>Generate 80 new chromosomes by mixing and matching the top 20 network’s chromosomes by performing simple random <a title="description of GA" href="http://en.wikipedia.org/wiki/Genetic_algorithm" target="_blank">crossover and mutation</a>.  The theory is that the ‘good stuff’ of how stuff should be done should be embedded in the chromosomes of the winners, and by combining the weights of the winners to generate new chromosomes, the hope is the ‘offsprings’ will also be good, or better than the parents.  From these 80 new chromosomes, generate 80 new networks.</p>
  </li>
  <li>
    <p>Repeat tasks (2) -&gt; (4) until the desired score of the best network is somewhat satisfactory.</p>
  </li>
</ol>

<div id="attachment_275" class="wp-caption alignnone">
  <a href="http://blog.otoro.net/wp-content/uploads/sites/2/2015/02/crossover.jpg"><img src="http://blog.otoro.net/wp-content/uploads/sites/2/2015/02/crossover.jpg"></a>

  <p class="wp-caption-text">
    crossover and mutation example
  </p>
</div>

<p>That’s the gist of the CNE algorithm.  It solves a big problem where we can easily define and score the goal, and not worry about backprop and what target values to train.  We can simply use the above algorithm to generate a motor speed for each time step and supply the network with current orientation/velocity states, and watch it search the solution space to find one that is satisfactory.  I like this method more in systems that contains many local maximums, so for SGD you just end up with the local maximum, while the NE algorithms has a better chance to find a much better local, or even global maximum.  In addition, it is relatively simple to incorporate recurrent neural network structures using CNE.</p>

<p>There are still many problems with CNE in the literature, where the algorithm actually loses diversity and converges also to a local maximum solution, and much of the work in this subfield is to find more advanced algorithms (in the thesis above, ESP and NEAT are named).</p>

<p>I will try to hack convnet.js and implement a simple CNE trainer that can train a network to achieve some score, and put it to test later to see if the pendulum problem may be attacked using these methods.</p>
</section>

    
  
  </div><br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via blog.otoro.net http://blog.otoro.net<br>
<br>
October 31, 2018 at 05:39PM
